### PR TITLE
Ensure that QCloseEvents can be vetoed by wx

### PR DESCRIPTION
**Pull request by matthew-griffin**
For some unknown reason, QCloseEvents are the only type of Qt event that should be **_ignored_** to indicate that they have been dealt with and the default close action should not occur (see https://doc.qt.io/archives/qq/qq11-events.html#acceptorignore). The QtHandleCloseEvent function also had its condition reversed as the Close function returns true to indicate that the window should be closed, ie. the event wasn't handled/vetoed.

I've also added a check on whether the window is enabled before allowing close, as a disabled Qt window does not disable the close button. This meant that disabled windows could be closed, which can currently be seen in the 'dialogs' sample if you try to close the main window while displaying a 'Busy info dialog'.

### CODE REVIEW

**Comment by vadz** [comment]
Thanks for the change and the explanations/link to the docs!

I don't really understand the meaning of `GetHandle() != handler` check, but shouldn't the return value be inverted in this case too?

**Comment by matthew-griffin** [comment]
I'm also unsure of the reason for the `GetHandle() != handler` check but I was wary of inverting it as that would mean that windows would not close in this situation when they previously did. Without an example of a legitimate reason for this happening, it seemed like a safety check where you wouldn't want to be locked into a window that wouldn't close.

I could go either way on changing `closeEvent()` instead of `QtHandleCloseEvent()`, it seemed to me like it was better for it to remain similar to the surrounding handler functions and not have multiple checks for whether to call `ignore()` on the event.

File: src/qt/window.cpp [hunk 1]
```diff
@@ -1490,7 +1490,10 @@ bool wxWindowQt::QtHandleCloseEvent ( QWidget *handler, QCloseEvent *WXUNUSED( e
     if ( GetHandle() != handler )
         return false;
 
-    return Close();
+    if (!IsEnabled())
+        return true;
```
Comment (catalinr) [inline src/qt/window.cpp]: Why should the window care if it was enabled or not? `Close()` can be called programmatically for a disabled window as well. Should it not close in that case?

File: src/qt/window.cpp [hunk 2]
```diff
@@ -1490,7 +1490,10 @@ bool wxWindowQt::QtHandleCloseEvent ( QWidget *handler, QCloseEvent *WXUNUSED( e
...
+
+    return !Close();
```
Comment (catalinr) [inline src/qt/window.cpp]: The names here are terrible :(
FWIW the meaning of "handle" in the name of this function is opposite to the meaning in the description of `wxWindow::Close()`. Maybe this function is worth a comment, something like `// Returns true if the closing should be vetoed and false if the window should be closed`.

File: src/qt/window.cpp (hunk 1 above)
Comment (matthew-griffin) [reply src/qt/window.cpp]: This was to restore the behaviour that was previously present in the MSW version, where clicking the close button of a disabled window would not close the window or generate a wx event.

Comment (catalinr) [reply src/qt/window.cpp]: You mean that right now, a disabled top level window will still react to a click on the [also disabled] close button?!

File: src/qt/window.cpp (hunk 2 above)
Comment (matthew-griffin) [reply src/qt/window.cpp]: I don't disagree that this section of the code is confusing, not sure where exactly the comment would be best placed. Perhaps the whole group of QtHandle... functions in qt/window.h should have a comment indicating that they return true to indicate that the event has been vetoed as they all work the same and can in theory be overridden by inheriting classes.

Comment (catalinr) [reply src/qt/window.cpp]: > not sure where exactly the comment would be best placed

I guess at the beginning of the function / right before it?

File: src/qt/window.cpp (hunk 1 above)
Comment (matthew-griffin) [reply src/qt/window.cpp]: Yes, I verified this in the 'dialogs' sample by opening: Dialogs -> Informative Dialogs -> Busy Info Dialog and then clicking the close button on the parent window. I guess the problem is more that Qt generates close events even when the window is disabled but I could not find any way to change this behaviour, other than dealing with the qt close event.

Comment (catalinr) [reply src/qt/window.cpp]: Amazing... oh well..
Can you please add a comment explaining this, maybe before the `if`?

Also, please correct the spaces for the same `if (  )`.

**Comment by vadz** [comment]
Thanks for the update! I've merged this into qt-fixed branch (#1552) and will merge it in master soon.

**Pull request closed without merging**